information/specification.md

```diff
@@ -0,0 +1,706 @@
+back to the wallet.
+
+Coin selection refers to the process of choosing _unspent outputs_ from a
+wallet's UTxO set in order to make one or more payments, and arranging for
+_change_ to be paid back to the wallet by creating one or more _change
+outputs_.
+
+## Why is Coin Selection Non-Trivial?
+
+### Issues
+
+There are a number of issues which make the problem of coin selection more
+complicated than it would initially appear.
+
+ * Each transaction has a _maximum size_, as defined by the protocol. The size
+   of a transaction increases as we add more inputs or outputs.
+
+   Therefore, there's a practical limit on the number of coins we can select
+   for any given transaction.
+
+ * The most obvious strategy for coin selection, which consists of trying to
+   get as close to the requested value as possible, will tend (over time) to
+   create a lot of _dust_: small unspent outputs.
+
+   Dust outputs are a problem, because even if the total value of dust in a
+   wallet is more than enough to cover a given target amount, if we attempt to
+   include that dust in a given transaction, we may run out of space (by
+   reaching the transaction size limit) before we can cover the target amount.
+
+ * The most obvious strategy for paying change, which consists of making a
+   single change output with the exact excess value, will tend (over time) to
+   reduce the size of the UTxO set. This is bad for two reasons:
+
+    1. Having a small UTxO set limits the number of payments that we can make
+       in parallel. Since a single UTxO entry can only be used to pay for a
+       single output, we need at least as many UTxO entries as there are
+       outputs.
+
+    2. The approach of coalescing all change into a single output is widely
+       considered to have negative privacy implications. The discussion of
+       privacy issues is beyond the scope of this article.
+
+### Goals
+
+There are several desirable properties that we would like a coin selection
+algorithm to have. These properties include:
+
+ * A coin selection algorithm should, over the course of time, aim to generate
+   and maintain a UTxO set with _useful_ outputs: that is, outputs that allow
+   us to process future payments with a _minimum_ number of inputs.
+
+ * A coin selection algorithm should employ strategies to limit the
+   amount of _dust_ that accumulates in the UTxO set.
+
+For more information on dust avoidance, see [Self Organisation in Coin
+Selection](#self-organization-in-coin-selection).
+
+# Definitions
+
+### Address
+
+An _address_ is a unique identifier that represents a destination for a payment.
+
+Addresses are owned (and generated) by individual wallets.
+
+In general, coin selection algorithms are agnostic to the choice of address
+format. Therefore, the details of individual address formats are not described
+within this document.
+
+### Coin Amount
+
+A _coin amount_ is a non-negative integer value that represents a number of
```

Comment:
It even is a _positive_ integer value. There's in principle, no such thing as a coin of value `0` (although quite practical in many calculations). This is true at least in Byron and I know there was discussion for Shelley about whether or not a transaction with an output of value `0` should be considered valid.

Reply:
> It even is a _positive_ integer value. There's in principle, no such thing as a coin of value `0` (although quite practical in many calculations). This is true at least in Byron and I know there was discussion for Shelley about whether or not a transaction with an output of value `0` should be considered valid.

Good catch. I initially wanted to include the possibility of zero, as this would allow us to talk about temporary amounts used in calculations (which occasionally end up as `0`, only to be filtered out later on). On reflection, I think it makes sense to exclude this case.

Fixed in 64e0dfc49e88f2ff1c3bd93a279b3a44d0d9ceec.